modules/grandpa/src/benchmarking.rs

```diff
@@ -0,0 +1,264 @@
+
+	}: submit_finality_proof(RawOrigin::Signed(caller), header, justification)
+	verify {
+		let header: BridgedHeader<T, I> = bp_test_utils::test_header(One::one());
+		let expected_hash = header.hash();
+
+		assert_eq!(<BestFinalized<T, I>>::get(), expected_hash);
+		assert!(<ImportedHeaders<T, I>>::contains_key(expected_hash));
+	}
+
+	// What we want to check here is the effect of many pre-commits on justification verification.
+	// We do this by creating many forks, whose head will be used as a signed pre-commit in the
+	// final justification.
+	submit_finality_proof_on_many_forks {
+		let p in 1..T::MaxBridgedValidatorCount::get();
+
+		let caller: T::AccountId = whitelisted_caller();
+
+		let authority_list = accounts(p as u8)
+			.iter()
+			.map(|id| (AuthorityId::from(*id), 1))
+			.collect::<Vec<_>>();
+
+		let init_data = InitializationData {
+			header: bp_test_utils::test_header(Zero::zero()),
+			authority_list,
+			set_id: TEST_GRANDPA_SET_ID,
+			is_halted: false,
+		};
+
+		initialize_bridge::<T, I>(init_data);
+		let header: BridgedHeader<T, I> = bp_test_utils::test_header(One::one());
+
+		let params = JustificationGeneratorParams {
+			header: header.clone(),
+			round: TEST_GRANDPA_ROUND,
+			set_id: TEST_GRANDPA_SET_ID,
+			authorities: accounts(p as u8).iter().map(|k| (*k, 1)).collect::<Vec<_>>(),
+			depth: 2,
+			forks: p,
+		};
+
+		let justification = make_justification_for_header(params).encode();
+
+	}: submit_finality_proof(RawOrigin::Signed(caller), header, justification)
+	verify {
+		let header: BridgedHeader<T, I> = bp_test_utils::test_header(One::one());
+		let expected_hash = header.hash();
+
+		assert_eq!(<BestFinalized<T, I>>::get(), expected_hash);
+		assert!(<ImportedHeaders<T, I>>::contains_key(expected_hash));
+	}
+
+	// Here we want to find out the overheaded of looking through consensus digests found in a
+	// header. As the number of logs in a header grows, how much more work do we require to look
+	// through them?
+	//
+	// Note that this should be the same for looking through scheduled changes and forces changes,
+	// which is why we only have one benchmark for this.
+	find_scheduled_change {
+		// Not really sure what a good bound for this is.
+		let n in 1..1000;
+
+		let mut logs = vec![];
+		for i in 0..n {
+			// We chose a non-consensus log on purpose since that way we have to look through all
+			// the logs in the header
+			logs.push(sp_runtime::DigestItem::Other(vec![]));
+		}
+
+		let mut header: BridgedHeader<T, I> = bp_test_utils::test_header(Zero::zero());
+		let digest = header.digest_mut();
+		*digest = sp_runtime::Digest {
+			logs,
+		};
+
+	}: {
+		crate::find_scheduled_change(&header)
+	}
+
+	// What we want to check here is how long it takes to read and write the authority set tracked
+	// by the pallet as the number of authorities grows.
+	read_write_authority_sets {
+		// The current max target number of validators on Polkadot/Kusama
+		let n in 1..1000;
```

Comment:
I don't quite get where you have got values:
```rust
	pub const RialtoValidatorCount: u32 = 255;
	pub const WestendValidatorCount: u32 = 255;
```
then :) And what's we're going to do with Kusama/Polkadot/Westend/Rococo benchmarks?

On Rialto max validators count is 5. Where did you get that value for Westend? I thought that params for all polka-like chains are similar?

Reply:
Right now I'm just using the max value allowed for this configuration option for all networks (see #846). Since the max value is above the actual validator number for all supported networks this is fine. The downside is that calls will end up being more expensive initially.

This might change a bit with the changes in #847 so I'll need to revisit this.

Reply:
Yes, but you're saying in the comment that for Polkadot/Kusama "max target number of validators" is 1000. So iiuc on Westend it is also 1000. Am I right here? Where did you got this thousand? We'e verifying justifications from Westend => we may overflow this 255 value for Westend justifications => end up with underpaying?

It is fine to keep it as is if you'll just remove these constants in next PRs.

Reply:
> Yes, but you're saying in the comment that for Polkadot/Kusama "max target number of validators" is 1000. So iiuc on Westend it is also 1000.

I don't think so actually. Since there's no incentive to run a validator on Westend nobody is doing it. From what I can tell there are only like 10 validators on Westend, with no "goal" of it reaching 1000.

> We'e verifying justifications from Westend => we may overflow this 255 value for Westend justifications => end up with underpaying?
>
> It is fine to keep it as is if you'll just remove these constants in next PRs.

If Westend had more than 255 validators, yes this would be a concern. However, it doesn't. At the same time I do plan to raise these limits (if we keep them) so that they are above the current target validator set size of 1000.